packages/design-system/src/components/TextField/TextInput.tsx

```diff
@@ -51,14 +56,18 @@ export interface TextInputProps {
   /**
    * Whether or not the text field is a multiline text field
    */
-  multiline?: boolean;
+  multiline?: MultilineValue;
   name?: string;
   /**
    * Sets `inputMode`, `type`, and `pattern` to improve accessiblity and consistency for number fields. Use this prop instead of `type="number"`, see [here](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) for more information.
    */
   numeric?: boolean;
-  onBlur?: (...args: any[]) => any;
-  onChange?: (...args: any[]) => any;
+  onBlur?: (
+    e: React.FocusEvent<MultilineValue extends true ? HTMLTextAreaElement : HTMLInputElement>
```

Comment:
Didn't realize you could use ternaries in a <Type>!

Reply:
I think TypeScript type annotations are Turing complete at this point 😆. Honestly it feels kind of icky, and there are probably several ways we could go about defining these types. I was spending too much time on this, though. We might want to go back at some point in the future and do another cleanup pass on the types (as long as the API doesn't change).